### PR TITLE
[SCTP] improve payload queue push performance

### PR DESCRIPTION
**Pull request by KillingSpark**
As discussed in #360 Gathering packets to send is a big chunk of the work the Association::write_loop is doing while in a critical section.

This PR improves the performance of this by making the payload queue more performant to push to.

Previously this did a full mergesort (O(n*log(n))) on all the in-flight TSNs, now it does a binary search (O(log(n)))

### CODE REVIEW

File: sctp/src/queue/payload_queue.rs [hunk 1]
```diff
@@ -24,26 +24,27 @@ impl PayloadQueue {
...
     }
 
     pub(crate) fn push_no_check(&mut self, p: ChunkPayloadData) {
+        let tsn = p.tsn;
         self.n_bytes += p.user_data.len();
-        self.sorted.push(p.tsn);
-        self.chunk_map.insert(p.tsn, p);
+        self.chunk_map.insert(tsn, p);
         self.length.fetch_add(1, Ordering::SeqCst);
-        self.update_sorted_keys();
```
Comment (melekes) [inline sctp/src/queue/payload_queue.rs:46]: 🤦‍♂️

File: sctp/src/queue/payload_queue.rs [hunk 2]
```diff
@@ -57,19 +58,14 @@ impl PayloadQueue {
...
-
+        self.push_no_check(p);
         true
     }
 
     /// pop pops only if the oldest chunk's TSN matches the given TSN.
     pub(crate) fn pop(&mut self, tsn: u32) -> Option<ChunkPayloadData> {
-        if !self.sorted.is_empty() && tsn == self.sorted[0] {
-            self.sorted.remove(0);
+        if Some(&tsn) == self.sorted.front() {
```
Comment (melekes) [inline sctp/src/queue/payload_queue.rs:67]:
```suggestion
        if Some(&tsn) == self.sorted.pop_front() {
```

File: sctp/src/queue/payload_queue.rs [hunk 3]
```diff
@@ -57,19 +58,14 @@ impl PayloadQueue {
...
+            self.sorted.pop_front();
```
Comment (melekes) [inline sctp/src/queue/payload_queue.rs:68]:
```suggestion
```

**Review by melekes — COMMENTED** [review]
❤️

**Comment by KillingSpark** [comment]
Sorry I must have made a mistake, the tests started failing. I'll fix it soon

File: sctp/src/queue/payload_queue.rs [hunk 4]
```diff
@@ -65,7 +75,7 @@ impl PayloadQueue {
     /// pop pops only if the oldest chunk's TSN matches the given TSN.
     pub(crate) fn pop(&mut self, tsn: u32) -> Option<ChunkPayloadData> {
         if Some(&tsn) == self.sorted.front() {
```
Comment (melekes) [inline sctp/src/queue/payload_queue.rs:77]: you can pop_front here. that way you won't need to do it later and would be able to remove assert_eq! completely.

Comment (KillingSpark) [reply sctp/src/queue/payload_queue.rs:77]: I don't think so. Doing so broke a few tests, and it wouldn't do what the signature says anymore.
Note that this is not binding `tsn` as part of a pattern but that this actually an equality check, and pop is only called if the tsns match. That tripped me up the first time I changed this part.

I should just remove the assert_eq! it doesn't do anything.

**Review by melekes — APPROVED** [review]

**Pull request merged**